pkg/serviceaccounts/oauthclient/oauthclientregistry.go

```diff
@@ -10,27 +12,88 @@ import (
 	"k8s.io/kubernetes/pkg/serviceaccount"
 
 	scopeauthorizer "github.com/openshift/origin/pkg/authorization/authorizer/scope"
+	osclient "github.com/openshift/origin/pkg/client"
 	oauthapi "github.com/openshift/origin/pkg/oauth/api"
 	"github.com/openshift/origin/pkg/oauth/registry/oauthclient"
+	routeapi "github.com/openshift/origin/pkg/route/api"
+	"k8s.io/kubernetes/pkg/util/sets"
 )
 
 const (
 	OAuthRedirectURISecretAnnotationPrefix = "serviceaccounts.openshift.io/oauth-redirecturi."
 	OAuthWantChallengesAnnotationPrefix    = "serviceaccounts.openshift.io/oauth-want-challenges"
+
+	OAuthRedirectModelAnnotationPrefix              = "serviceaccounts.openshift.io/oauth-redirectmodel."
+	OAuthRedirectModelAnnotationURIPrefix           = OAuthRedirectModelAnnotationPrefix + "uri."
+	OAuthRedirectModelAnnotationResourcePrefix      = OAuthRedirectModelAnnotationPrefix + "resource."
+	OAuthRedirectModelAnnotationURISchemePrefix     = OAuthRedirectModelAnnotationURIPrefix + "scheme."
+	OAuthRedirectModelAnnotationURIPortPrefix       = OAuthRedirectModelAnnotationURIPrefix + "port."
+	OAuthRedirectModelAnnotationURIPathPrefix       = OAuthRedirectModelAnnotationURIPrefix + "path."
+	OAuthRedirectModelAnnotationResourceKindPrefix  = OAuthRedirectModelAnnotationResourcePrefix + "kind."
+	OAuthRedirectModelAnnotationResourceNamePrefix  = OAuthRedirectModelAnnotationResourcePrefix + "name."
+	OAuthRedirectModelAnnotationResourceGroupPrefix = OAuthRedirectModelAnnotationResourcePrefix + "group."
+
+	RouteKind = "Route"
 )
 
+var invalidRedirectURI = errors.New("Invalid redirect URI")
+
 type saOAuthClientAdapter struct {
 	saClient     kclient.ServiceAccountsNamespacer
 	secretClient kclient.SecretsNamespacer
+	routeClient  osclient.RoutesNamespacer
 
 	delegate    oauthclient.Getter
 	grantMethod oauthapi.GrantHandlerType
 }
 
+type model struct {
+	scheme string
+	port   string
+	path   string
+
+	group string
+	kind  string
+	name  string
+}
+
+type redirectURI struct {
+	scheme string
+	host   string
+	port   string
+	path   string
+}
+
+func (uri *redirectURI) getURI() (string, error) {
+	var host string
+	if uri.port != "" {
+		host = fmt.Sprintf("%s:%s", uri.host, uri.port)
+	} else {
+		host = uri.host
+	}
+	u, err := url.Parse(fmt.Sprintf("%s://%s/%s", uri.scheme, host, uri.path))
+	if err == nil && u.Fragment == "" {
+		return u.String(), nil
+	}
+	return "", invalidRedirectURI
+}
+
+func (uri *redirectURI) merge(m model) {
+	if m.scheme != "" {
+		uri.scheme = m.scheme
+	}
+	if m.path != "" {
+		uri.path = m.path
+	}
+	if m.port != "" {
+		uri.port = m.port
+	}
```

Comment:
I populate the defaults / known values in `updateRouteMap` - haven't figured out how to tell from the route data if the route wants to use TLS or not.